rules/models/rollup.go

```diff
@@ -0,0 +1,229 @@
+
+// RollupRuleViews belonging to a ruleset indexed by uuid.
+// Each value contains the entire snapshot history of the rule.
+type RollupRuleViews map[string][]*RollupRuleView
+
+// NewRollupTarget takes a RollupTargetView and returns the equivalent RollupTarget.
+func NewRollupTarget(t RollupTargetView) RollupTarget {
+	return RollupTarget(t)
+}
+
+// ToRollupTargetView returns the equivalent ToRollupTargetView.
+func (t RollupTarget) ToRollupTargetView() RollupTargetView {
+	return RollupTargetView(t)
+}
+
+// Sort sorts the policies inside the rollup target.
+func (t *RollupTarget) Sort() {
+	sort.Strings(t.Tags)
+	sort.Sort(policy.ByResolutionAscRetentionDesc(t.Policies))
+}
+
+// Equals determines whether two rollup targets are equal.
+func (t *RollupTarget) Equals(other *RollupTarget) bool {
+	if t == nil && other == nil {
+		return true
+	}
+	if t == nil || other == nil {
+		return false
+	}
+	if t.Name != other.Name {
+		return false
+	}
+	if len(t.Tags) != len(other.Tags) {
+		return false
+	}
+	for i := 0; i < len(t.Tags); i++ {
+		if t.Tags[i] != other.Tags[i] {
+			return false
+		}
+	}
+	return policy.Policies(t.Policies).Equals(policy.Policies(other.Policies))
+}
+
+// NewRollupRule takes a RollupRuleView and returns the equivalent RollupRule.
+func NewRollupRule(rrv *RollupRuleView) RollupRule {
+	targets := make([]RollupTarget, len(rrv.Targets))
+	for i, t := range rrv.Targets {
+		targets[i] = NewRollupTarget(t)
+	}
+	return RollupRule{
+		ID:                  rrv.ID,
+		Name:                rrv.Name,
+		Filter:              rrv.Filter,
+		Targets:             targets,
+		CutoverMillis:       rrv.CutoverNanos / nanosPerMilli,
+		LastUpdatedBy:       rrv.LastUpdatedBy,
+		LastUpdatedAtMillis: rrv.LastUpdatedAtNanos / nanosPerMilli,
+	}
+}
+
+// ToRollupRuleView returns the equivalent ToRollupRuleView.
+func (r RollupRule) ToRollupRuleView() *RollupRuleView {
+	targets := make([]RollupTargetView, len(r.Targets))
+	for i, t := range r.Targets {
+		targets[i] = t.ToRollupTargetView()
+	}
+
+	return &RollupRuleView{
+		ID:      r.ID,
+		Name:    r.Name,
+		Filter:  r.Filter,
+		Targets: targets,
+	}
+}
+
+// Equals determines whether two rollup rules are equal.
+func (r *RollupRule) Equals(other *RollupRule) bool {
+	if r == nil && other == nil {
+		return true
+	}
+	if r == nil || other == nil {
+		return false
+	}
+	return r.Name == other.Name &&
+		r.Filter == other.Filter &&
+		rollupTargets(r.Targets).Equals(other.Targets)
+}
+
+// Sort sorts the rollup targets inside the rollup rule.
+func (r *RollupRule) Sort() {
+	for i := range r.Targets {
+		r.Targets[i].Sort()
+	}
+	sort.Sort(rollupTargetsByNameTagsAsc(r.Targets))
+}
+
+// SameTransform returns whether two rollup targets have the same transformation.
+func (rtv *RollupTargetView) SameTransform(other RollupTargetView) bool {
+	if rtv.Name != other.Name {
+		return false
+	}
+	if len(rtv.Tags) != len(other.Tags) {
+		return false
+	}
+	clonedTags := rtv.Tags
```

Comment:
This doesn't actually make a copy of the tags, so it'll actually mutate the tags. I think we need to clone the tags here since `SameTransform` is not supposed to change the state of `RollupTargetView`.